tests/Normalization.hs

```diff
@@ -1,18 +1,73 @@
-{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE OverloadedLists #-}
+{-# LANGUAGE OverloadedStrings#-}
 module Normalization (normalizationTests) where
 
 import           Dhall.Core
+import           Dhall.TypeCheck (typeOf, X(..))
 import           Test.Tasty
 import           Test.Tasty.HUnit
 
 normalizationTests :: TestTree
-normalizationTests = testGroup "normalization" [ constantFolding ]
+normalizationTests = testGroup "normalization" [ constantFolding
+                                               , fusion
+                                               ]
 
 constantFolding :: TestTree
-constantFolding = testGroup "folding of constants" [ naturalPlus ]
+constantFolding = testGroup "folding of constants" [ naturalPlus, optionalFold, optionalBuild ]
 
 naturalPlus :: TestTree
-naturalPlus = testCase "natural plus" $ normalize' (NaturalPlus (NaturalLit 1) (NaturalLit 2)) @?= NaturalLit 3
+naturalPlus = testCase "natural plus" $ do
+  isNormalized e @?= False
+  normalize' e @?= NaturalLit 3
+  where e = NaturalPlus (NaturalLit 1) (NaturalLit 2)
 
-normalize' :: Expr () () -> Expr () ()
+optionalFold :: TestTree
+optionalFold = testGroup "Optional/fold" [ just, nothing ]
+  where test label inp out = testCase label $ do
+             isNormalized (e inp) @?= False
+             normalize' (e inp) @?= out
+        e inp = (OptionalFold `App` Text `App` OptionalLit Text inp `App`
```

Comment:
Clever use of infix `App`!  That would make the `normalization` code a lot easier to read with fewer parentheses since you I believe can also pattern match on an infix constructor